### PR TITLE
More clear tests

### PR DESCRIPTION
**Pull request by virgil-serbanuta**

### CODE REVIEW

File: ulm-semantics/test/execution.md [hunk 1]
```diff
@@ -63,68 +69,71 @@ module ULM-TEST-EXECUTION
...
+    syntax Identifier ::= "append_bytes_raw"  [token]
+                        | "buffer_id"  [token]
+                        | "bytes_hooks"  [token]
+                        | "empty"  [token]
+                        | "str"  [token]
+
+    syntax EncodeValues ::= encodeArgsToEncodeValues(EncodeArgs)  [function, total]
+    rule encodeArgsToEncodeValues(.EncodeArgs) => .EncodeValues
+    rule encodeArgsToEncodeValues(E:Expression : T:Type , Es:EncodeArgs)
+        => E : T , encodeArgsToEncodeValues(Es)
```
Comment (sskeirik) [inline ulm-semantics/test/execution.md:94]: It looks like `EncodeArg` and `EncodeValue` have the same syntax. Is this the intended behavior? At the moment I don't fully understand how these sorts are being used.

File: ulm-semantics/test/execution.md [hunk 2]
```diff
@@ -199,6 +208,100 @@ module ULM-TEST-EXECUTION
...
+            )
+    rule encodeValues(Statements:NonEmptyStatements)
+        =>  encodeValues({.InnerAttributes Statements buffer_id })
+    rule encodeValues(ptrValue(_, u64(BytesId))) => encodeValues(ulmBytesId(BytesId))
+    rule encodeValues(ulmBytesValue(B:Bytes)) => encodedValues(B)
+
+    rule SetAccountStorageHook(evaluatedStorageKey(B:Bytes), ptrValue(_, Value:Value))
+        => #SetAccountStorageHook
+                ( Bytes2Int(Keccak256raw(B), BE, Unsigned)
+                , valueToInteger(Value)
```
Comment (sskeirik) [inline ulm-semantics/test/execution.md:268]: It seems that, in the case `valueToInteger` returns error, the rules below on 270 and 274 will not cover all possible cases and evaluation will get stuck.

File: ulm-semantics/test/execution.md [hunk 3]
```diff
@@ -199,6 +208,100 @@ module ULM-TEST-EXECUTION
...
+    rule GetAccountStorageHook(evaluatedStorageKey(B:Bytes))
+        => #GetAccountStorageHook(Bytes2Int(Keccak256raw(B), BE, Unsigned))
+    rule mock #GetAccountStorageHook(Key:Int) Result:UlmHookResult
+        => mock
+            GetAccountStorageHook(Key)
+            Result
+    rule list_mock #GetAccountStorageHook(Key:Int) Result:UlmHookResult
+        => list_mock
+            GetAccountStorageHook(Key)
+            Result
```
Comment (sskeirik) [inline ulm-semantics/test/execution.md:288]: The use of overloaded names `Set/GetAccountStorageHook` and `#Set/GetAccountStorageHook` both in this file and in `ulm-semantics/main/hooks/ulm.md` makes the logic a bit hard to follow in these rules.

It looks in this file the declared sharp versions are intermediate forms --- perhaps they could be renamed to avoid conflict with the overloaded version in `ulm-semantics/main/hooks/ulm.md`?

I understand that the non-sharp version in this file is designed to make the test files look like the underlying functions in the semantics, so the overloading there seems to have a potentially bigger payoff in test readability.

**Review by sskeirik — APPROVED** [review]
Looks good overall. I'm approving with some questions and comments about minor possible improvements. Feel free to merge once you've responded and/or addressed the questions/comments.

File: ulm-semantics/test/execution.md [hunk 4]
```diff
@@ -63,68 +69,71 @@ module ULM-TEST-EXECUTION
...
+        => concat
+            (   let buffer_id = :: bytes_hooks :: empty( .CallParamsList );
+                .NonEmptyStatements
+            ,   codegenValuesEncoder(buffer_id, encodeArgsToEncodeValues(Args))
+            )
+
+    rule encode_call_data C:EncodeCall
+        => encodeCallData(extractCallSignature(C), encodeInstructions(C))
+
+    syntax ExecutionItem ::= encodeCallData(BytesOrError, NonEmptyStatementsOrError)
```
[the remark this answers is not in the document]
Comment (virgil-serbanuta) [reply ulm-semantics/test/execution.md:110]: It can't have a "total" attribute because it's not defined when either the first or the second argument is an error. Since partial functions have all sorts of problems (e.g. having a `Left => Right` rewrite rule in a semantics where `Right` may be undefined makes the entire semantics unsound, or whatever is the proper word for it), I would prefer to use them only when absolutely needed.

Even if we ignore that, I would argue that not making this a function makes the code more debugable. When evaluating an undefined function, the LLVM backend will print the function that it can't evaluate, which is sort of nice. However, it's harder to get the entire context in which this happened (i.e., the entire configuration). If `encodeCallData` is not a function, you get the entire configuration.

With the Haskell backend, if we make this a function, it depends. If there is a `#Ceil` definition for this, then proving would just succeed (since bottom implies everything), although that's certainly not what we want here (although I think that they may have added a warning for this case, or something like that). Without a `#Ceil` definition, you'd get the full configuration with an unevaluated function, which is fine for debugging.

File: ulm-semantics/test/execution.md (hunk 2 above)
Comment (virgil-serbanuta) [reply ulm-semantics/test/execution.md:268]: Yes, it gets stuck, and valueToInteger evaluates to a term containing an error message somewhat explaining what went wrong. That's by design. I mean, is there a better option if the user asks to add a mock for a `SetAccountStorage(key, non-integer-value)` call?

Of course, one option would be to provide an `Int` directly to `SetAccountStorageHook ` instead of an `Expression`, and I'll do that if you think it's better, but this PR attempted to make tests more readable, and I'm assuming that using `Expression` helps.

File: ulm-semantics/test/execution.md (hunk 3 above)
Comment (virgil-serbanuta) [reply ulm-semantics/test/execution.md]: Renamed.

File: ulm-semantics/test/execution.md [hunk 5]
```diff
@@ -63,68 +69,71 @@ module ULM-TEST-EXECUTION
...
+        </k>
+        <test-stack>
+            ListItem(ptrValue(_, u64(V))) => .List
             ...
-         </k> [owise]
+        </test-stack>
+        <ulm-bytes-buffers>
+            Buffers:Map
+        </ulm-bytes-buffers>
+        requires B:Bytes:KItem ==K Buffers[MInt2Unsigned(V)] orDefault 0
```
[the remark this answers is not in the document]
Comment (virgil-serbanuta) [reply ulm-semantics/test/execution.md:132]: Yes. I added a comment saying that explicitly.

File: ulm-semantics/test/execution.md (hunk 1 above)
Comment (virgil-serbanuta) [reply ulm-semantics/test/execution.md]: They are mostly the same thing. There are some obscure reasons for keeping them separate (mostly I was hesitant about using `EncodeValues` for parsing), but I think that you're probably right and I should merge them. See the last commit of this PR. If you decide that, for some reasons, they should be separate, I can create a new PR undoing that commit.
